internal/ghinstance/host.go

```diff
@@ -36,7 +37,18 @@ func IsTenancy(h string) bool {
 // reports whether it found the tenant name.
 func TenantName(h string) (string, bool) {
 	normalizedHostName := NormalizeHostname(h)
-	return cutSuffix(normalizedHostName, "."+tenancyHost)
+	t, f := cutSuffix(normalizedHostName, "."+tenancyHost)
+
+	if !f {
+		return t, f
+	}
+
+	// make sure tenant name is valid
+	re := regexp.MustCompile(`^[a-z0-9\-]+$`)
+	if !re.MatchString(t) {
+		return "", false
+	}
+	return t, true
```

Comment:
@kommendorkapten : Could you elaborate on how `gh` would behave when invalid tenant hosts were specified?  Any console output you could share?

Depending on ☝️, we might need to revisit this because this is a larger issue than just core `gh` as [`cli/go-gh`](https://github.com/cli/go-gh/blob/trunk/pkg/auth/auth.go) has similar capability. 🤔

Reply:
That is a good question. Currently some commands are failing before this test happens, as they are instantiating the API client first, which tries to parse (and possibly connect) the URL. But as this function can be called from everywhere we should be on par in `cli/go-gh`, I can prepare a PR once we have a rough agreement on how the cli should behave. 

As the tenant is derived from the hostname, this PR simply prompts `invalid hostname $host`. Having all commands exhibit a similar behaviour should be good. 

One thing though, `TenantName` returns bool (`found`), but with a corresponding check of `IsTenancy`, an error case can be detected.

Reply:
@andyfeller any thoughts?

Reply:
I'm trying to think of a middle ground as this feels suboptimal. 🤔  Let me see if I can unwind what's going on here to see what alternatives might be available.

## Is there somewhere else we can handle invalid tenancy names?

The only place I see [`ghrepo.TenantName` being called](https://github.com/search?q=repo%3Acli%2Fcli+%22.TenantName%22&type=code) is within `ghrepo.FormatRemoteUrl`, which is [used throughout the code base](https://github.com/search?q=repo%3Acli%2Fcli+%22ghrepo.FormatRemoteUrl%22&type=code):

https://github.com/cli/cli/blob/78c1d00eccac1b2ae82ac0bfeea3e2292c98056a/internal/ghrepo/repo.go#L95-L103

Above, you state there are some scenarios that are failing, however I don't have a clear idea of concrete examples:

> Currently some commands are failing before this test happens, as they are instantiating the API client first, which tries to parse (and possibly connect) the URL.

Putting that aside, I also see elsewhere in the code we have `HostnameValidator` which is also [used throughout code](https://github.com/search?q=repo%3Acli/cli%20HostnameValidator&type=code):

https://github.com/cli/cli/blob/78c1d00eccac1b2ae82ac0bfeea3e2292c98056a/internal/ghinstance/host.go#L62C6-L70

**Should this be extended to see if a provided host is a tenant and throw an error if invalid instead?**

Reply:
created https://github.com/cli/cli/issues/9598 for tracking.